### DISCUSSION TITLE
Add @deprecated annotation to GDScript

### DISCUSSION THREAD

**Discussion by blackears** [godotengine/godot-proposals#6559] [Scripting]
When you start working with larger projects, it can be useful to mark obsolete code that is still being used with a @deprecated tag to alert other programmers that they should no longer be using a class or method.  It would be nice if this could be added to GDScript.

**Comment by theraot** [comment]
I would also find this useful. I would want it to:

1. cause a warning
2. allow me to write a custom message for said warning

**Comment by AThousandShips** [comment]
Would be nice to integrate it with the engine too and make bound functions of built in types indicate the same, the documentation system has support for deprecation of methods, constants, properties, and classes

**Reply by dalexeev** [reply]
It's already there for classes:
![](https://user-images.githubusercontent.com/47700418/227728018-d1776b6c-b3d2-4f63-9ae9-93fe9faa0aee.png)
![](https://user-images.githubusercontent.com/47700418/227728062-5e77cda1-5eea-44c7-869a-760635e0fe79.png)

For methods and properties (when calling a setter/getter) there are macros `WARN_DEPRECATED` and `WARN_DEPRECATED_MSG`. For GDScript, you can use `push_warning()` in the same manner (but the warning will only be at runtime).

Annotations have the advantage that a warning will be displayed in the editor if the type can be inferred. But I think this information should also be available to the core API so that the GDScript analyzer can display warnings for native properties as well.

**Reply by AThousandShips** [reply]
Would be great, I think the deprecation and experimental tags are a bit cumbersome as they are only accessible via the doc files, making it hard to notice in the source, adding the functionality to `ClassDB` would help

**Reply by AThousandShips** [reply]
Been experimenting with integrating the experimental and deprecated properties into `ClassDB`, will see how it goes but promising so far
